src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,549 @@
+                // 2. Add using of new namespace (for each reference's container).
+                // 3. Try to simplify qualified names introduced from step(1).
+
+                var addImportService = document.GetLanguageService<IAddImportsService>();
+                var syncNamespaceService = document.GetLanguageService<ISyncNamespaceService>();
+                
+                var (documentWithRefFixed, containers) = 
+                    await FixReferencesAsync(document, addImportService, syncNamespaceService, refLocations, _newNamespaceParts, cancellationToken)
+                        .ConfigureAwait(false);
+
+                var optionSet = await documentWithRefFixed.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var placeSystemNamespaceFirst = optionSet.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, documentWithRefFixed.Project.Language);
+
+                var documentWithAdditionalImports = await AddImportsInContainersAsync(
+                    documentWithRefFixed, 
+                    addImportService,
+                    containers,
+                    ImmutableArray.Create(_state.TargetNamespace), 
+                    placeSystemNamespaceFirst, 
+                    cancellationToken).ConfigureAwait(false);
+
+                // Need to invoke formatter explicitly since we are doing the diff merge ourselves.
+                var formattedDocument = await Formatter.FormatAsync(documentWithAdditionalImports, Formatter.Annotation, optionSet, cancellationToken)
+                    .ConfigureAwait(false);
+
+                return await Simplifier.ReduceAsync(formattedDocument, optionSet, cancellationToken).ConfigureAwait(false);
+            }
+
+            /// <summary>
+            /// Fix each reference and return a collection of proper containers (innermost container
+            /// with imports) that new import should be added to based on reference locations.
+            /// If <paramref name="namespaceParts"/> is specified (not default), the fix would be:
+            ///     1. qualify the reference with new namespace and mark it for simplification, or
+            ///     2. find and mark the qualified reference for simplification.
+            /// Otherwise, there would be no namespace replacement.
+            /// </summary>
+            private async Task<(Document, ImmutableArray<SyntaxNode>)> FixReferencesAsync(
+                Document document, 
+                IAddImportsService addImportService, 
+                ISyncNamespaceService syncNamespaceService,
+                IEnumerable<ReferenceLocation> refLocations, 
+                ImmutableArray<string> namespaceParts,
+                CancellationToken cancellationToken)
+            {
+                var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+                var root = editor.OriginalRoot;
+                var containers = new HashSet<SyntaxNode>();
```

Comment:
if you are going to use PooledHashSet below, you could use it here as well.

Reply:
Fixed